### PR TITLE
PS-71 QR Scanner

### PR DESCRIPTION
**Pull request by laniszew**

### CODE REVIEW

File: src/global/Components/QrScanner/QrScanner.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
+import React, { useEffect, useCallback } from 'react';
+import styled from 'styled-components';
+import Webcam from 'react-webcam';
+import { QRCode } from 'jsqr';
+import getPixels from 'get-pixels';
+/* eslint-disable */
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/QrScanner.tsx]: why we disable eslint here?

File: src/global/Components/QrScanner/QrScanner.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  border-radius: 50%;
+  display: flex;
+  position: absolute;
+  bottom: 30px;
+  left: 50%;
+  transform: translate(-50%);
+`;
+
+type ExtendedWebcam = {
+  ref: React.RefObject<Webcam>;
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/QrScanner.tsx]: You can add RefObject into import at the top of this file and use just RefObject insted of React.RefObject

File: src/global/Components/QrScanner/QrScanner.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+type defaultProps = {
+  videoConstraints: MediaTrackConstraints;
+  interval: number;
+  buttonMode: boolean;
+  className: string;
+  onError: (error: string) => void;
+};
+
+type Props = {
+  onCode: (code: QRCode) => void;
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/QrScanner.tsx]: maybe this methods can be optional so we can have one type QrScannerProps ?

File: src/global/Components/QrScanner/QrScanner.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  onUserMediaError?: (error: string) => void;
+} & defaultProps;
+
+type ImagePxels = {
+  data: Uint8ClampedArray;
+  offset: number;
+  shape: Array<number>;
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/QrScanner.tsx]: You have no eslint warning here? We can go with just `number[]`.

File: src/global/Components/QrScanner/QrScanner.tsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    webWorker.addEventListener('message', onMessage);
+    const timer = setInterval(capture, interval);
+    return () => {
+      if (webWorker) {
+        webWorker.terminate();
+        webWorker = null;
+      }
+      clearInterval(timer);
+    };
+  }, [capture, interval]);
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/QrScanner.tsx]: Maybe we can create custom hook `useQrScannerWebWorker` for this code to clarify  render stuff?

File: src/global/Components/QrScanner/QrScanner.tsx [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        onUserMediaError={onUserMediaError}
+      />
+      {buttonMode ? <Button onClick={capture} /> : null}
+    </Wrapper>
+  );
+};
+
+QrScanner.defaultProps = {
+  videoConstraints: { facingMode: 'environment' },
+  interval: 500,
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/QrScanner.tsx]: magic number

File: src/global/Components/QrScanner/QrScanner.tsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  onError: (err: string) => {
+    console.error(err);
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/QrScanner.tsx]: we want to console that?

File: src/global/Components/QrScanner/decodeWorker.worker.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import jsQr from 'jsqr';
+
+const ctx: Worker = self as any;
+
+ctx.addEventListener('message', ({ data }) => {
+  const scannedQr = jsQr(data.data, data.shape[0], data.shape[1]);
+  if (scannedQr) {
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/decodeWorker.worker.ts]: maybe one line space here?

File: src/pages/Home/Home.tsx [hunk 9]
```diff
@@ -1,3 +1,10 @@
 import React, { FunctionComponent } from 'react';
+import QrScanner from '../../global/Components/QrScanner/QrScanner';
```
Comment (lmazurowski) [inline src/pages/Home/Home.tsx]: we should define aliases for this stuff, but that's not a part of that PR :)

File: src/global/Components/QrScanner/decodeWorker.worker.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import jsQr from 'jsqr';
+
+const ctx: Worker = self as any;
+
+ctx.addEventListener('message', ({ data }) => {
```
Comment (lmazurowski) [inline src/global/Components/QrScanner/decodeWorker.worker.ts]: what is a type of callback argument?

File: src/global/Components/QrScanner/QrScanner.tsx (hunk 1 above)
Comment (laniszew) [reply src/global/Components/QrScanner/QrScanner.tsx]: Cause of 
@typescript-eslint/ban-ts-ignore
Changed to disable-next-line

File: src/global/Components/QrScanner/QrScanner.tsx (hunk 3 above)
Comment (laniszew) [reply src/global/Components/QrScanner/QrScanner.tsx]: It doesn't make too much sense to be optional, as we wouldn't use scanning then

File: src/global/Components/QrScanner/QrScanner.tsx (hunk 2 above)
Comment (laniszew) [reply src/global/Components/QrScanner/QrScanner.tsx]: done

File: src/global/Components/QrScanner/QrScanner.tsx (hunk 7 above)
Comment (laniszew) [reply src/global/Components/QrScanner/QrScanner.tsx]: I think so, that's default if no onError callback was provided

**Review by pumanitro — APPROVED** [review]

**Pull request merged**